### PR TITLE
Added test case for the wait function of write handler.

### PR DESCRIPTION
**Pull request by grv07**
This change is [Reviewable](https://reviewable.io/reviews/jonhoo/left-right/90)

### CODE REVIEW

**Comment by grv07** [comment]
Hello,
I have added a test case for the wait function of the write handler.

This test checks a case where one of the readers is still reading and the hander calls the wait fn.
I try to simulate the read handle behaviour by creating a new thread that will increase the epoch by one when scheduled.
Thanks.

**Comment by jonhoo** [comment]
Hmm, I don't think this test does quite what you want it to. In particular, the thread that increments may end up running before `wait` is even called, in which case `wait` won't wait at all. There isn't _really_ a good way to fix that problem except by creating a `loom` test to test both interleavings, which I know is something @mlodato517 was looking at a little while back.

**Comment by grv07** [comment]
Yes, I understand your point.
Apart from loom do you think park_timeout will be good fit for this problem?
If yes should I update it with park_timeout ?

**Comment by jonhoo** [comment]
`park_timeout` could work too! You'd probably want to first `park_timeout` _without_ the reader allowing the writer to make progress, check that the writer did in fact not make progress, then `park`, then progress the reader, and then check that the writer was then unblocked.

File: src/write.rs [hunk 1]
```diff
@@ -450,6 +465,48 @@ mod tests {
...
+        w.last_epochs = vec![2, 2, 1];
+        let mut epochs_slab = Slab::new();
+        epochs_slab.insert(Arc::new(AtomicUsize::new(2)));
+        epochs_slab.insert(Arc::new(AtomicUsize::new(2)));
+        epochs_slab.insert(Arc::clone(&holded_epoch));
+
+        // A new thread act as a reader thread that will increase epoch by 1 and break wait cycle.
+        let move_holded_epoch = Arc::clone(&holded_epoch);
+        let progress_handle = thread::spawn(move || {
+            thread::park_timeout(Duration::from_secs(2));
```
Comment (jonhoo) [inline src/write.rs:492]: Ah, if this is what you wanted to do, you'll probably want to use `sync::Barrier` instead so that you don't end up with a race condition in the test :)

File: src/write.rs [hunk 2]
```diff
@@ -179,6 +187,10 @@ where
                     continue 'retry;
                 }
             }
+            #[cfg(test)]
```
Comment (jonhoo) [inline src/write.rs:190]: Maybe stick this below the loop rather than just before the `break`?

File: src/write.rs [hunk 3]
```diff
@@ -450,6 +465,48 @@ mod tests {
...
+    #[test]
+    fn wait_test() {
+        use std::thread;
+        use std::time::Duration;
+        let (mut w, _r) = crate::new::<i32, _>();
+
+        // Case 1: If epoch is set to default.
+        let test_epochs: crate::Epochs = Default::default();
+        let mut test_epochs = test_epochs.lock().unwrap();
+        w.wait(&mut test_epochs);
```
Comment (jonhoo) [inline src/write.rs:477]: It'd be good to document here why you expect `wait` to return immediately.

File: src/write.rs [hunk 4]
```diff
@@ -450,6 +465,48 @@ mod tests {
...
+
+        // Case 2: If one of the reader is still reading(epoch is odd and count is same as in last_epoch)
+        // and wait has been called.
+        let holded_epoch = Arc::new(AtomicUsize::new(1));
```
Comment (jonhoo) [inline src/write.rs:481]:
```suggestion
        let held_epoch = Arc::new(AtomicUsize::new(1));
```

File: src/write.rs (hunk 1 above)
Comment (jonhoo) [reply src/write.rs:492]: I think this may also end up nicer if you make the spawned thread do the call to `wait` instead. Then the main test code can first check that it's _not_ waiting before the barrier, then check that it _is_ waiting shortly after, then check that it stays that way after a call to `yield_now`, then bump the `held_epoch`, and then check that it can join the spawned thread.

File: src/write.rs [hunk 5]
```diff
@@ -450,6 +465,66 @@ mod tests {
...
+
+        w.last_epochs = vec![2, 2, 1];
+        let mut epochs_slab = Slab::new();
+        epochs_slab.insert(Arc::new(AtomicUsize::new(2)));
+        epochs_slab.insert(Arc::new(AtomicUsize::new(2)));
+        epochs_slab.insert(Arc::clone(&held_epoch));
+
+        let barrier = Arc::new(Barrier::new(2));
+
+        let is_waiting = Arc::clone(&w.is_waiting);
```
Comment (jonhoo) [inline src/write.rs:492]: You can probably just assert that this is `false` here rather than doing it in both threads before the barrier wait below. Arguably you don't even need that pre-assert.

File: src/write.rs [hunk 6]
```diff
@@ -450,6 +465,66 @@ mod tests {
...
+        let c = Arc::clone(&barrier);
```
Comment (jonhoo) [inline src/write.rs:493]: I'd call this `barrier2` or something like it so it's clearer below that the thread is also waiting on the barrier. Single-character variable names tend to make code hard to follow.

File: src/write.rs [hunk 7]
```diff
@@ -450,6 +465,66 @@ mod tests {
...
+
+        let m_is_waiting = Arc::clone(&w.is_waiting);
+        let test_epochs = Arc::new(Mutex::new(epochs_slab));
+
+        let wait_handle = thread::spawn(move || {
+            let is_waiting = m_is_waiting.load(Ordering::Relaxed);
+            assert_eq!(false, is_waiting);
+            let mut test_epochs = test_epochs.lock().unwrap();
```
Comment (jonhoo) [inline src/write.rs:501]: Let's take this lock after the barrier wait instead of before. Feels wrong to hold a lock while waiting on a barrier.

File: src/write.rs [hunk 8]
```diff
@@ -450,6 +465,66 @@ mod tests {
...
+        let is_waiting_v = is_waiting.load(Ordering::Relaxed);
+        assert_eq!(false, is_waiting_v);
+
+        barrier.wait();
+        // make sure that writer wait() will call first then change the held epoch.
+        thread::yield_now();
+
+        let is_waiting_v = is_waiting.load(Ordering::Relaxed);
+        // check if is_waiting is set to true after calling wait.
+        assert_eq!(true, is_waiting_v);
```
Comment (jonhoo) [inline src/write.rs:515]: I don't think you're guaranteed that this is `true` even after a call to `yield_now`. You'll probably instead want `yield_now` in a loop until `is_waiting` is `true`. And then yield once more and check that it's still true.

File: src/write.rs [hunk 9]
```diff
@@ -450,6 +465,66 @@ mod tests {
...
+
+        held_epoch.fetch_add(1, Ordering::SeqCst);
+
+        let _ = wait_handle.join();
```
Comment (jonhoo) [inline src/write.rs:519]: A comment here saying why we now expect the wait to have made progress would be good.

File: src/write.rs [hunk 10]
```diff
@@ -128,16 +130,22 @@ where
...
     }
 
-    fn wait(&mut self, epochs: &mut MutexGuard<'_, slab::Slab<Arc<atomic::AtomicUsize>>>) {
+    fn wait(&mut self, epochs: &mut MutexGuard<'_, slab::Slab<Arc<AtomicUsize>>>) {
         let mut iter = 0;
         let mut starti = 0;
 
+        #[cfg(test)]
+        {
+            self.is_waiting.swap(true, Ordering::Relaxed);
```
Comment (jonhoo) [inline src/write.rs:147]: I believe this can just be a `store`?

File: src/write.rs [hunk 11]
```diff
@@ -181,6 +189,10 @@ where
             }
             break;
         }
+        #[cfg(test)]
+        {
+            self.is_waiting.swap(false, Ordering::Relaxed);
```
Comment (jonhoo) [inline src/write.rs:194]: Same here — I think this can be a `store`

File: src/write.rs [hunk 12]
```diff
@@ -450,6 +465,73 @@ mod tests {
...
+        let wait_handle = thread::spawn(move || {
+            barrier2.wait();
+            let mut test_epochs = test_epochs.lock().unwrap();
+            w.wait(&mut test_epochs);
+        });
+
+        barrier.wait();
+
+        // make sure that writer wait() will call first only then update the held epoch.
+        'retry: loop {
```
Comment (jonhoo) [inline src/write.rs:509]: I wonder, can this just be a `while` loop?
```rust
while !is_waiting.load(Ordering::Relaxed) {
    thread::yield_now();
}
```

File: src/write.rs [hunk 13]
```diff
@@ -450,6 +465,73 @@ mod tests {
...
+            let is_waiting_v = is_waiting.load(Ordering::Relaxed);
+            if is_waiting_v {
+                thread::yield_now();
+                break 'retry;
+            }
+        }
+
+        let is_waiting_v = is_waiting.load(Ordering::Relaxed);
+        // check if is_waiting is set to true after calling wait.
+        assert_eq!(true, is_waiting_v);
```
Comment (jonhoo) [inline src/write.rs:520]: I think you can remove this — you wouldn't get to this `assert` unless `is_waiting_v` is already true.

File: src/write.rs [hunk 14]
```diff
@@ -450,6 +465,73 @@ mod tests {
...
+
+        held_epoch.fetch_add(1, Ordering::SeqCst);
+
+        // join to make sure that wait fn must return after the progress/increment
+        // of held_epoch.
+        let _ = wait_handle.join();
+
+        let count = held_epoch.load(Ordering::Relaxed);
+        assert_eq!(count, 2);
```
Comment (jonhoo) [inline src/write.rs:529]: What is this assertion checking?

File: src/write.rs [hunk 15]
```diff
@@ -505,29 +505,19 @@ mod tests {
...
-            if is_waiting_v {
-                thread::yield_now();
-                break 'retry;
-            }
         }
-
-        let is_waiting_v = is_waiting.load(Ordering::Relaxed);
-        // check if is_waiting is set to true after calling wait.
-        assert_eq!(true, is_waiting_v);
+        thread::yield_now();
```
Comment (jonhoo) [inline src/write.rs:512]: Why is this extra yield needed?

**Comment by jonhoo** [comment]
I think we're almost there now!

File: src/write.rs [hunk 16]
```diff
@@ -450,6 +465,62 @@ mod tests {
...
+            thread::yield_now();
+        }
+
+        held_epoch.fetch_add(1, Ordering::SeqCst);
+
+        // join to make sure that wait must return after the progress/increment
+        // of held_epoch.
+        let _ = wait_handle.join();
+
+        // check that writes waiting state is set to false after wait returns.
```
Comment (jonhoo) [inline src/write.rs:519]: I think I'd just remove this assertion — it doesn't really add much, and anything that's in the test that's not relevant _to_ the test might confuse later readers :)

**Comment by jonhoo** [comment]
Oh, and if you can merge from `master`, we should get CI testing and coverage analysis for this PR too!

**Review by jonhoo — APPROVED** [review]

**Comment by jonhoo** [comment]
Awesome, thank you for sticking with it!
